####################################################################################################

ptm <- proc.time()

# ... loading the TSRchitect library:
library(TSRchitect)

BAMDIR  <- c("/path/to/RAMPAGE/output/alignments>")
DmAnnot  <- c("/path/to//Drosophila_melanogaster.BDGP5.78.gff")

# ... initializing the tssObj object:
DmRAMPAGE <- loadTSSobj(experimentTitle="RAMPAGE Tutorial", inputDir=BAMDIR, inputType="bam", isPairedEnd=TRUE, sampleNames=c("E1h","E2h", "E3h", "L1", "L2", "L3"), replicateIDs=c(1,1,1,2,2,2)) ## 

# ... converting BAM data into TSS information and attaching it to the tssObj object:
DmRAMPAGE <- bamToTSS(DmRAMPAGE)

# ... constructing the tag count per TSS data matrices:
DmRAMPAGE <- processTSS(experimentName=DmRAMPAGE, n.cores=4, tssSet="all", writeTable=TRUE)

# ... identifying TSRs (promoters):
DmRAMPAGE <- determineTSR(experimentName=DmRAMPAGE, n.cores=4, tsrSetType="replicates", tssSet="all", tagCountThreshold=25, clustDist=20, writeTable=TRUE)

# ... now adding tag count output:
#
DmRAMPAGE <- addTagCountsToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=1, tagCountThreshold=10, writeTable=TRUE)
DmRAMPAGE <- addTagCountsToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=2, tagCountThreshold=10, writeTable=TRUE)
DmRAMPAGE <- addTagCountsToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=3, tagCountThreshold=10, writeTable=TRUE)
DmRAMPAGE <- addTagCountsToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=4, tagCountThreshold=10, writeTable=TRUE)
DmRAMPAGE <- addTagCountsToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=5, tagCountThreshold=10, writeTable=TRUE)
DmRAMPAGE <- addTagCountsToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=6, tagCountThreshold=10, writeTable=TRUE)

# ... now adding annotation to the replicates:
#
DmRAMPAGE <- importAnnotationExternal(experimentName=DmRAMPAGE, fileType="gff3", annotFile=DmAnnot)
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=1, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=2, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=3, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=4, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=5, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=6, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)

# ... merging the samples into embryonic and larval
#
DmRAMPAGE <- mergeSampleData(DmRAMPAGE)

# ... identifying TSRs (promoters) from the merged samples:
DmRAMPAGE <- determineTSR(experimentName=DmRAMPAGE, n.cores=4, tsrSetType="merged", tssSet="all", tagCountThreshold=40, clustDist=20, writeTable=TRUE)

# ... now adding annotation to the merged samples:
#
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="merged", tsrSet=1, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)
DmRAMPAGE <- addAnnotationToTSR(experimentName=DmRAMPAGE, tsrSetType="merged", tsrSet=2, upstreamDist=1000, downstreamDist=200, feature="gene", featureColumnID="ID", writeTable=TRUE)

# Optional:  Let's explore how many TSS are precisely identified:
#
#system("mv TSRset-1.txt TSRset-1.txtORIG")

# Optional:  Write bed-formatted output:
#
#writeTSR(experimentName=DmRAMPAGE, tsrSetType="replicates", tsrSet=1, fileType="bed")

proc.time() - ptm

# ... uncomment the following line if you would like to the data generated by the above
#     for future reloading with the R load command:
save(DmRAMPAGE, file="DmRAMPAGE.RData")
